### PR TITLE
Reentrancy Counter: Update supported chains

### PR DESCRIPTION
**Pull request by areezladhani**
# reentrancy-counter

## update supported chains

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] New Bot
- [ ] Fix a bot error
- [x] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

- Updated the bot to support the following networks: Ethereum, Optimism, Binance Smart Chain, Polygon, Fantom, Arbitrum, Avalanche.

### CODE REVIEW

**Comment by Vxatz** [comment]
Hey @areezladhani, can you also add the repository in the package.json? (https://github.com/NethermindEth/Forta-Agents/tree/main/reentrancy-counter)

File: reentrancy-counter/package.json [hunk 1]
```diff
@@ -3,6 +3,15 @@
   "version": "0.0.1",
```
Comment (Vxatz) [inline reentrancy-counter/package.json:3]:
```suggestion
  "version": "0.0.2",
```

File: reentrancy-counter/package.json [hunk 2]
```diff
@@ -3,6 +3,15 @@
...
   "agentId": "nethforta-25",
   "description": "Forta Agent Typescript starter project",
```
Comment (Vxatz) [inline reentrancy-counter/package.json:5]:
```suggestion
  "description": "Reentrancy Monitoring",
```

**Review by Vxatz — CHANGES_REQUESTED** [review]

**Review by 0xtaf — CHANGES_REQUESTED** [review]
hey Areez,

could you please update forta-agent version? this version is quite old and there are some vulnerable dependencies inside this one.

after updating the package, can you run `npm install` again to see if there are vulnerable packages? if there are, try running `npm audit fix`

again after updating the package, the code will throw an error for `Receipt`, inside testing. to prevent that, I would suggest using `forta-agent-tools` package and fixing the test cases. (we now have `TestTransactionEvent` instead of `createTransactionEvent`). you'll understand the usage after checking some of our projects. 

it's gonna be a good exercise to learn about our `gam`

**Comment by areezladhani** [comment]
Hi, i've updated the forta-agent version to deal with all vulnerabilities and updated the test file accordingly.  There is three commits in total, the first one contains the majority of the changes. The last two are just small changes i missed the first time around.

File: reentrancy-counter/package.json [hunk 3]
```diff
@@ -17,7 +27,8 @@
   },
   "dependencies": {
     "bignumber.js": "^9.0.1",
-    "forta-agent": "^0.0.15"
+    "forta-agent": "^0.1.15",
+    "forta-agent-tools": "^3.1.2"
```
Comment (0xtaf) [inline reentrancy-counter/package.json:31]: since this is only used for testing, it should be inside `devDependencies`

**Review by 0xtaf — CHANGES_REQUESTED** [review]
great work updating the tests and forta version! 
there's just 1 small detail I want to mention below.

File: reentrancy-counter/README.md [hunk 4]
```diff
@@ -2,19 +2,24 @@
...
 
-This agent detects report transaction where reentrancy ocur, based on the calls stack provided in the transaction traces.
-It report the amount of repeated calls with different severities according to different thresholds.
+This bot checks transactions for reentrancy. This is done by the bot watching the call stack in transaction traces. It then reports the number of recurrent calls with varying severity based on various thresholds.
 
 ## Supported Chains
 
 - Ethereum
+- Optimism
+- Binance Smart Chain
```
Comment (0xtaf) [inline reentrancy-counter/README.md:11]:
```suggestion
- BNB Smart Chain
```
I did the same one but this is what they use :)

File: reentrancy-counter/README.md [hunk 5]
```diff
@@ -2,19 +2,24 @@
...
 - NETHFORTA-25
-  - Fired when in a transaction ocur multiples nested calls to the same contract (reentrancy)
-  - It report all the posible severities based on different thresholds of the amount of calls
+  - Fired when multiple nested calls occur to the same contract in a transaction (Reentrancy)
+  - It reports all possible severities based on different call volume thresholds.
   - Type is always set to "suspicious"
-  - The metadata contains the reentered address & the reentrancy counter
+  - The metada contains the following information:
+    - The contract address where the reentrancy occurred
+    - A reentrancy counter based on how many times it occurred
```
Comment (0xtaf) [inline reentrancy-counter/README.md:25]:
```suggestion
  - The metadata includes:
    - `address`: The contract address where the reentrancy occurred
    - `reentrancyCount`: A reentrancy counter based on how many times it occurred
```
this is our current format now

**Comment by areezladhani** [comment]
Hi @0xtaf, @Vxatz

I've implemented the latest changes.

**Review by Vxatz — APPROVED** [review]
LGTM ✅

File: reentrancy-counter/package.json [hunk 6]
```diff
@@ -1,8 +1,18 @@
...
+  "chainIds": [
+    1,
+    10,
+    56,
+    137,
+    250,
+    42161,
+    43114
+  ],
   "scripts": {
```
Comment (0xtaf) [inline reentrancy-counter/package.json:16]: can you replace the scripts with this object:

```  
"scripts": {
    "start": "npm run start:dev",
    "start:dev": "nodemon --watch src --watch forta.config.json -e js,json --exec \"forta-agent run\"",
    "start:prod": "forta-agent run --prod",
    "tx": "forta-agent run --tx",
    "block": "forta-agent run --block",
    "range": "forta-agent run --range",
    "file": "forta-agent run --file",
    "publish": "forta-agent publish",
    "info": "forta-agent info",
    "logs": "forta-agent logs",
    "push": "forta-agent push",
    "disable": "forta-agent disable",
    "enable": "forta-agent enable",
    "keyfile": "forta-agent keyfile",
    "test": "jest"
  },
```

**Review by 0xtaf — CHANGES_REQUESTED** [review]
hey @areezladhani ,

sorry to request changes one by one but I realize something new.
since the Forta is updated from a quite old version, scripts and Dockerfile are also changed,

I wrote the scripts in the comment but I need to type content of Dockerfile here, can you replace your Dockerfile with this?

```
FROM node:12-alpine AS builder
WORKDIR /app
COPY . .
RUN npm ci
RUN npm run build

# Final stage: copy compiled Javascript from previous stage and install production dependencies
FROM node:12-alpine
ENV NODE_ENV=production
# Uncomment the following line to enable agent logging
# LABEL "network.forta.settings.agent-logs.enable"="true"
WORKDIR /app
COPY --from=builder /app/dist ./src
COPY package*.json ./
RUN npm ci --production
CMD [ "npm", "run", "start:prod" ]
```

**Review by 0xtaf — APPROVED** [review]
LGTM 🚀

**Pull request merged**